File: aeperion-evdb/generate-tree.py
import json
import os
import re

ECOSYS = os.path.join(os.path.dirname(__file__), 'ecosys')

def extract_meta(filepath):
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            lines = f.readlines()
        title = (lines[0] or '').replace('#', '').strip() if lines else ''
        category = ''
        level = ''
        desc = ''
        for i in range(min(len(lines), 30)):
            l = lines[i]
            if l.startswith('**Categoría:**'): category = l.replace('**Categoría:**', '').strip()
            if l.startswith('**Nivel de Autonomía:**'): level = l.replace('**Nivel de Autonomía:**', '').strip()
            if l.startswith('**Nivel de Complejidad:**'): level = l.replace('**Nivel de Complejidad:**', '').strip()
            if l.startswith('**Tipo:**'): category = l.replace('**Tipo:**', '').strip()
            if l.startswith('**Compatibilidad:**'): level = l.replace('**Compatibilidad:**', '').strip()
        for i in range(1, min(len(lines), 25)):
            l = lines[i].strip()
            if not l or l.startswith('---') or l.startswith('```') or re.match(r'^##\s+\d', l) or l.startswith('- ') or l.startswith('"'):
                continue
            cleaned = l.replace('**', '').strip()
            if cleaned and len(cleaned) > 25 and 'Categoría' not in cleaned and 'Nivel de' not in cleaned and 'Tipo:' not in cleaned and 'Compatibilidad' not in cleaned and 'skill_id' not in cleaned and '"version"' not in cleaned:
                desc = cleaned[:200]
                break
        return {'title': title, 'category': category, 'level': level, 'desc': desc}
    except:
        return {'title': '', 'category': '', 'level': '', 'desc': ''}

def readable_name(name):
    name = re.sub(r'\.(md|txt)$', '', name)
    name = re.sub(r'^Prompt_', '', name)
    name = re.sub(r'^Skill_', '', name)
    name = re.sub(r'^Agente_', '', name)
    name = name.replace('_', ' ')
    name = re.sub(r'([a-z])([A-Z])', r'\1 \2', name)
    name = re.sub(r'\b(De|Del|La)\b', lambda m: m.group(1).lower(), name)
    return name

def scan_dir(dirpath, parent_id=''):
    try:
        entries = sorted(os.listdir(dirpath))
    except:
        return []
    entries = [e for e in entries if e != 'builder.py']
    groups = []
    prompts = []
    for entry in entries:
        full_path = os.path.join(dirpath, entry)
        rel_path = '/' + os.path.relpath(full_path, os.path.dirname(__file__)).replace('\\', '/')
        if os.path.isdir(full_path):
            children = scan_dir(full_path, entry)
            if children:
                node_id = (parent_id + '-' + entry).lower()
                node_id = re.sub(r'[^a-z0-9-]', '-', node_id)
                node_id = re.sub(r'-+', '-', node_id)
                groups.append({
                    'id': node_id,
                    'label': readable_name(entry),
                    'type': 'group',
                    'count': len(children),
                    'children': children,
                })
        elif entry.endswith('.md') or entry.endswith('.txt'):
            meta = extract_meta(full_path) if entry.endswith('.md') else {'title': '', 'category': '', 'level': '', 'desc': ''}
            with open(full_path, 'r', encoding='utf-8') as f:
                content = f.read()
            node_id = (parent_id + '-' + entry).lower()
            node_id = re.sub(r'[^a-z0-9-]', '-', node_id)
            node_id = re.sub(r'-+', '-', node_id)
            prompts.append({
                'id': node_id,
                'label': readable_name(entry),
                'type': 'prompt',
                'file': entry,
                'prompt': rel_path,
                'abspath': full_path,
                'content': content,
                'meta': meta,
            })
    if groups and prompts:
        node_id = (parent_id + '-general').lower()
        node_id = re.sub(r'[^a-z0-9-]', '-', node_id)
        node_id = re.sub(r'-+', '-', node_id)
        groups.append({
            'id': node_id,
            'label': 'General',
            'type': 'group',
            'count': len(prompts),
            'children': prompts,
        })
        groups.sort(key=lambda x: x['label'])
        return groups
    nodes = groups + prompts
    nodes.sort(key=lambda x: (0 if x['type'] == 'group' else 1, x['label']))
    return nodes

folders = ['Agentes', 'Prompts', 'Skills']
result = {}
for f in folders:
    dir_path = os.path.join(ECOSYS, f)
    if os.path.isdir(dir_path):
        result[f] = scan_dir(dir_path, f)

extra_dir = os.path.join(ECOSYS, 'Agents')
if os.path.isdir(extra_dir):
    extra = scan_dir(extra_dir, 'Agentes')
    if extra:
        result['Agentes'].append({
            'id': 'agentes-standalone',
            'label': 'Standalone Agents',
            'type': 'group',
            'count': len(extra),
            'children': extra,
        })

json_str = json.dumps(result, ensure_ascii=False)
outdir = os.path.dirname(__file__)
with open(os.path.join(outdir, 'tree-data.json'), 'w', encoding='utf-8') as f:
    json.dump(result, f, ensure_ascii=False, indent=2)
with open(os.path.join(outdir, 'tree-data.js'), 'w', encoding='utf-8') as f:
    f.write('window.__TREE__ = ' + json_str + ';')

def count_prompts(nodes):
    c = 0
    for n in nodes:
        if n['type'] == 'prompt':
            c += 1
        if 'children' in n:
            c += count_prompts(n['children'])
    return c

print('tree-data.json + tree-data.js regenerated')
for f in folders:
    items = result.get(f, [])
    total = count_prompts(items)
    print(f'  {f}: {len(items)} groups, {total} prompts')

File: aeperion-evdb/generate-tree.js
const fs = require('fs')
const path = require('path')

const ECOSYS = path.join(__dirname, 'ecosys')

function extractMeta(filePath) {
  try {
    const content = fs.readFileSync(filePath, 'utf-8')
    const lines = content.split('\n')
    const title = (lines[0] || '').replace(/^#\s+/, '').trim()

    let category = ''
    let level = ''
    let desc = ''

    for (let i = 0; i < Math.min(lines.length, 30); i++) {
      const l = lines[i]
      if (l.startsWith('**Categoría:**')) category = l.replace('**Categoría:**', '').trim()
      if (l.startsWith('**Nivel de Autonomía:**')) level = l.replace('**Nivel de Autonomía:**', '').trim()
      if (l.startsWith('**Nivel de Complejidad:**')) level = l.replace('**Nivel de Complejidad:**', '').trim()
      if (l.startsWith('**Tipo:**')) category = l.replace('**Tipo:**', '').trim()
      if (l.startsWith('**Compatibilidad:**')) level = l.replace('**Compatibilidad:**', '').trim()
    }

    for (let i = 1; i < Math.min(lines.length, 25); i++) {
      const l = lines[i].trim()
      if (!l || l.startsWith('---') || l.startsWith('```') || l.match(/^##\s+\d/) || l.startsWith('- ') || l.startsWith('"')) continue
      const cleaned = l.replace(/\*\*/g, '').trim()
      if (cleaned && cleaned.length > 25 && !cleaned.includes('Categoría') && !cleaned.includes('Nivel de') && !cleaned.includes('Tipo:') && !cleaned.includes('Compatibilidad') && !cleaned.includes('skill_id') && !cleaned.includes('"version"')) {
        desc = cleaned.substring(0, 200)
        break
      }
    }

    return { title, category, level, desc }
  } catch {
    return { title: '', category: '', level: '', desc: '' }
  }
}

function readableName(file) {
  let name = file.replace(/\.md$/, '').replace(/\.txt$/, '')
  name = name.replace(/^Prompt_/, '')
  name = name.replace(/^Skill_/, '')
  name = name.replace(/^Agente_/, '')
  name = name.replace(/_/g, ' ')
  name = name.replace(/([a-z])([A-Z])/g, '$1 $2')
  name = name.replace(/De /g, 'de ').replace(/Del /g, 'del ').replace(/La /g, 'la ')
  return name
}

function scanDir(dir, parentId = '') {
  const entries = fs.readdirSync(dir, { withFileTypes: true }).filter(e => e.name !== 'builder.py')
  const groups = []
  const prompts = []

  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name)
    const relPath = '/' + path.relative(__dirname, fullPath).replace(/\\/g, '/')

    if (entry.isDirectory()) {
      const children = scanDir(fullPath, entry.name)
      if (children.length > 0) {
        groups.push({
          id: (parentId + '-' + entry.name).replace(/[^a-zA-Z0-9-]/g, '-').toLowerCase().replace(/-+/g, '-'),
          label: readableName(entry.name),
          type: 'group',
          count: children.length,
          children,
        })
      }
    } else if (entry.isFile() && (entry.name.endsWith('.md') || entry.name.endsWith('.txt'))) {
      const meta = entry.name.endsWith('.md') ? extractMeta(fullPath) : { title: '', category: '', level: '', desc: '' }
      const content = fs.readFileSync(fullPath, 'utf-8')
      prompts.push({
        id: (parentId + '-' + entry.name).replace(/[^a-zA-Z0-9-]/g, '-').toLowerCase().replace(/-+/g, '-'),
        label: readableName(entry.name),
        type: 'prompt',
        file: entry.name,
        prompt: relPath,
        abspath: fullPath,
        content,
        meta,
      })
    }
  }

  // If there are both groups and orphan files, wrap files in a "General" group
  if (groups.length > 0 && prompts.length > 0) {
    groups.push({
      id: (parentId + '-general').replace(/[^a-zA-Z0-9-]/g, '-').toLowerCase().replace(/-+/g, '-'),
      label: 'General',
      type: 'group',
      count: prompts.length,
      children: prompts,
    })
    groups.sort((a, b) => a.label.localeCompare(b.label))
    return groups
  }

  // Otherwise return combined (if only files, return as-is)
  const nodes = [...groups, ...prompts]
  nodes.sort((a, b) => {
    if (a.type === 'group' && b.type !== 'group') return -1
    if (a.type !== 'group' && b.type === 'group') return 1
    return a.label.localeCompare(b.label)
  })
  return nodes
}

const folders = ['Agentes', 'Prompts', 'Skills']
const result = {}
for (const f of folders) {
  const dirPath = path.join(ECOSYS, f)
  if (fs.existsSync(dirPath)) {
    result[f] = scanDir(dirPath, f)
  }
}

// Include extra agent files (e.g. ecosys/Agents/) under Agentes as a General group
const extraAgentDir = path.join(ECOSYS, 'Agents')
if (fs.existsSync(extraAgentDir)) {
  const extra = scanDir(extraAgentDir, 'Agentes')
  if (extra.length > 0) {
    result['Agentes'].push({
      id: 'agentes-standalone',
      label: 'Standalone Agents',
      type: 'group',
      count: extra.length,
      children: extra,
    })
  }
}

const json = JSON.stringify(result)
fs.writeFileSync(path.join(__dirname, 'tree-data.json'), JSON.stringify(result, null, 2))
fs.writeFileSync(path.join(__dirname, 'tree-data.js'), 'window.__TREE__ = ' + json + ';')

function countPrompts(nodes) {
  let c = 0
  for (const n of nodes) {
    if (n.type === 'prompt') c++
    if (n.children) c += countPrompts(n.children)
  }
  return c
}

console.log('tree-data.json + tree-data.js regenerated')
for (const f of folders) {
  const items = result[f] || []
  const total = countPrompts(items)
  console.log(`  ${f}: ${items.length} groups, ${total} prompts`)
}
